### ISSUE TITLE
Bug in Accord.Statistics.Analysis.DistributionAnalysis

### ISSUE THREAD

**Issue by Burkhardus** [accord-net/framework#301]
When I use a certain set of values, the distribution analysis compute is not finishing but hanging for infinite time. This is causing the application to stop working without any information why. 

For my understanding there should be an error message if the data is not acceptable for computation or if an error occurs during computation. 

Furthermore I don't understand why the problem happening only on some kind of data and on others not.

Imports Accord.Statistics
Imports Accord.Statistics.Analysis
Imports Accord.Statistics.Distributions
 
Public Class Form1
 
    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
 
        Dim Working() As Double = {8.833, 8.385, 9.583, 9.002, 8.057, 9.052, 8.373, 10.768, 7.298, 7.701, 9.785, 7.606, 7.964, 8.914}
 
        Dim da1 As New DistributionAnalysis(Working)
        da1.Compute() ' here it works like intended
        MsgBox(da1.GoodnessOfFit(0).Name & " : " & da1.GoodnessOfFit(0).Distribution.ToString, vbOKOnly, "Result")
 
        Dim NotWorking() As Double = {1.003, 1.012, 1.011, 1.057, 1.033, 1.051, 1.045, 1.045, 1.037, 1.059, 1.028, 1.032, 1.029, 1.031, 1.029, 1.023, 1.035}
 
        Dim da2 As New DistributionAnalysis(NotWorking)
        da2.Compute() ' Here it hangs for no specific reason
        MsgBox(da2.GoodnessOfFit(0).Name & " : " & da2.GoodnessOfFit(0).Distribution.ToString, vbOKOnly, "Result")
 
    End Sub
 
End Class

**Comment by Burkhardus** [comment]
I analyzed the bug and I found that the root cause is in the fit method of the gamma distribution.

The while loop is never exited, because newK and oldK will never be the same. 

But this happens not always, but only for some very specific observation values. I am not expert enough for the underlying mathematics, so I believe I can not do much more than pointing roughly to the line of code, where the problem occurs.

```c#
public override void Fit(double[] observations, double[] weights, IFittingOptions options)
        {
            if (immutable)
                throw new InvalidOperationException("This object can not be modified.");

            if (options != null)
                throw new ArgumentException("This method does not accept fitting options.");

            if (weights != null)
                throw new ArgumentException("This distribution does not support weighted samples.");

            // Method from Choi, S.C.; Wette, R. (1969) "Maximum Likelihood Estimation 
            // of the Parameters of the Gamma Distribution and Their Bias", Technometrics,
            // 11(4) 683–690

            double lnsum = 0;
            for (int i = 0; i < observations.Length; i++)
                lnsum += Math.Log(observations[i]);

            double mean = observations.Mean();

            double s = Math.Log(mean) - lnsum / observations.Length;

            if (Double.IsNaN(s))
                throw new ArgumentException("Observation vector contains negative values.", "observations");

            // initial approximation
            double newK = (3 - s + Math.Sqrt((s - 3) * (s - 3) + 24 * s)) / (12 * s);

            // Use Newton-Raphson update
            double oldK;

            do
            {
                oldK = newK;
                double num = Math.Log(newK) - Gamma.Digamma(newK) - s;
                double den = (1 / newK) - Gamma.Trigamma(newK);
                newK = oldK - num / den;
            }
            while (!oldK.IsEqual(newK, 1e-10));

            double theta = mean / newK;

            init(theta, newK);
        }
```

**Comment by cesarsouza** [comment]
Hi Burkhardus,

Thank you for reporting the issue. I will include your example in the unit tests and investigate. In the meantime, a possible fix would be to add a 'max iterations' option for the Gamma fitting method, which at least would prevent the application from hanging. 

If you need a urgent fix, and if you currently have the sources of the library and you can change it temporarily, it should be possible to add a check for a limit of iterations within the do-while block inside the Gamma fit method.

Otherwise, I should be able to commit a fix in a few days.

**Comment by Burkhardus** [comment]
I have temporarily fixed the issue by lowering the tolerance value in the isEqual-function from 1e-10 to 1e-7, but I dont know if this is allowed, because I do not know enough about this kind of mathematics.

But I was reading about the Newton-Raphson procedure and that it is not converging in all cases. There are optimized versions available.

[http://scicomp.stackexchange.com/questions/30/why-is-newtons-method-not-converging](url)

[http://www.cs.sun.ac.za/~rw315/notes/newton-raphson.pdf](url)

**Comment by cesarsouza** [comment]
A fix will be committed in a few minutes.

**Comment by Burkhardus** [comment]
Hello Cesar,

with the new update, the Newton-Raphson method in the Gamma Distribution is now finishing after e.g. 1000 iterations, but this doesnt mean that it has aproximated the right results.

I have the feeling that this Newton-Raphson method did not fit the right parameters, because the density function gives wrong results (NaN) or -1#.IND on the given observation values, even if it found the Gama Distribition the most likely Distribution of the five.

I was reading about the N-R-method and I found that it sometimes has some problems to converge, The sequence can become oscillating or diverging. It is too complex for myself, as I am not a mathematician.

http://scicomp.stackexchange.com/questions/30/why-is-newtons-method-not-converging

 http://www.cs.sun.ac.za/~rw315/notes/newton-raphson.pdf

**Comment by Burkhardus** [comment]
I have furthermore analyzed the bug and I would like to provide more information on it. Please take a look at my test program and the screenshot. You can see, that the ppl created from the PDF contains no valid data for the y values. 

``` c#
namespace WindowsFormsApplication1
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {

            Double[] Dataset1 = { 8.833, 8.385, 9.583, 9.002, 8.057, 9.052, 8.373, 10.768, 7.298, 7.701, 9.785, 7.606, 7.964, 8.914 };
            Double[] Dataset2 = { 0.999, 0.996, 1.066, 1.041, 1.052, 1.043, 1.052, 1.046, 1.032, 1.038, 1.052, 1.038, 1.040, 1.056, 1.043, 1.044, 1.058, 1.044, 1.053, 1.046, 1.045, 1.046, 1.048, 1.051, 1.056, 1.058, 1.065, 1.051, 1.049, 1.049, 1.043, 1.042, 1.049, 1.042, 1.033, 1.042, 1.028, 1.031, 1.036, 1.044, 1.037, 1.054, 1.046, 1.061, 1.048, 1.048, 1.028, 1.030, 1.049, 1.038, 1.042, 1.057, 1.028, 1.047, 1.044, 1.040, 1.038, 1.044, 1.046, 1.035, 1.040, 1.039, 1.046, 1.050, 1.033, 1.032, 1.040, 1.040, 1.032, 1.045, 1.054, 1.040, 1.045, 1.045, 1.043, 1.066, 1.057, 1.052, 1.060, 1.069, 1.046, 1.041, 1.048, 1.057, 1.059, 1.076, 1.051, 1.057, 1.067, 1.062, 1.025, 1.072, 1.061, 1.067, 1.062, 1.055, 1.065, 1.047, 1.046, 1.047, 1.045, 1.038, 1.056, 1.037, 1.042, 1.044, 1.047, 1.037, 1.047, 1.041, 1.042, 1.042, 1.042, 1.040, 1.060, 1.066, 1.067, 1.077, 1.059, 1.067 };
            
        var da = new DistributionAnalysis();
            da.Learn(Dataset2);
        var result = da.GoodnessOfFit[0].Distribution.ToString();

        var hist = new Histogram();  
            hist.Compute(Dataset2);

        var DistGamma = new GammaDistribution();
            DistGamma.Fit(Dataset2);

        var ppl = new PointPairList();

            double x = hist.Range.Min;
            double stepx = (hist.Range.Max - hist.Range.Min) / 100;
            double y;

 
![gammadistributionfitandpdf1](https://cloud.githubusercontent.com/assets/22381218/18811655/3859e510-82ea-11e6-84bd-ce1c0cbed331.JPG)
           for (int i = 0; i < 100; i++)
            {
                y = DistGamma.ProbabilityDensityFunction(x);
                x = x + stepx;
                ppl.Add(x, y);
            }

        }
    }
}
```

**Comment by Burkhardus** [comment]
Hello Cesar,

I have tested your fix and it works now. But I found another issue in the Probability Density Function of the Gamma Distribution. I opened another Change Requestion on this bug.

I have compared the fitting results of your Newton-Raphson method with the results of Maximum Likelihood method and they are mostly smilar, approx 3 % different.

I would like to close this change requestion. Thank you for your fast reply.
